subprojects/core/src/main/java/org/gradle/execution/plan/InputAccessHierarchy.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.specs.Spec;
+import org.gradle.execution.plan.ValuedVfsHierarchy.ValueVisitor;
+import org.gradle.internal.collect.PersistentList;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.io.File;
+import java.util.function.Supplier;
+
+public class InputAccessHierarchy {
+    private volatile ValuedVfsHierarchy<ChildAccess> root;
+    private final Stat stat;
+
+    public InputAccessHierarchy(CaseSensitivity caseSensitivity, Stat stat) {
+        this.root = new ValuedVfsHierarchy<>(PersistentList.of(), EmptyChildMap.getInstance(), caseSensitivity);
+        this.stat = stat;
+    }
+
+    /**
+     * Returns all nodes which access the location.
+     *
+     * That includes node which access ancestors or children of the location.
+     */
+    public boolean isInput(String location) {
+        AccessVisitor visitor = new AccessVisitor();
+        VfsRelativePath relativePath = VfsRelativePath.of(location);
+        if (relativePath.length() == 0) {
```

Comment:
This feels like it shouldn't test the path length, but instead call something like `isEmpty()` or `isRoot()`.

It seems that `length()` should probably be package-private, and instead we should add a public method for `String.substring(VfsRelativePath.length() + 1)` which is the other use of this method outside the package.

Reply:
I tried doing something in this direction here: https://github.com/gradle/gradle/pull/19617

Is that what you had in mind?